unityfiles/Assets/Scripts/Game/Enemy/Behaviour/EnemyBehaviour.cs

```diff
@@ -0,0 +1,213 @@
+using System;
+using UnityEngine;
+
+/// <summary>
+/// Enemy AI logic. Controls enemies idle, patrol, chase and attack states.
+/// 
+/// </summary>
+[RequireComponent(typeof(Animator))]
+[RequireComponent(typeof(BoxCollider2D))]
+public class EnemyBehaviour : MonoBehaviour {
+
+	private enum BehaviourState {
+		IDLE,
+		PATROL,
+		CHASE,
+		ATTACK
+	}
+
+	[SerializeField]
+	private bool visualizeVision = true;
+
+	private readonly string WALK_TRIGGER_NAME = "walk";
+	private readonly string CHASE_TRIGGER_NAME = "chase";
+	private readonly string ATTACK_TRIGGER_NAME = "attack";
+
+	private readonly int MAX_PATROLTIME = 5;
+	private readonly int MIN_PATROLTIME = 3;
+
+	private readonly int MAX_IDLE_TIME = 4;
+	private readonly int MIN_IDLE_TIME = 2;
+	private readonly float TIME_BETWEEN_ATTACKS = 1f;
+	private readonly float ATTACK_DISTANCE = 1.5f;
+
+	private BehaviourState CurrentState { get; set; }
+
+	public float PatrolSpeed { get; set; }
+	public float ChaseSpeed { get; set; }
+	private float VisionLength { get; set; }
+	private float PatrolTime { get; set; }
+	private float IdleTime { get; set; }
+	private float AttackTimer { get; set; }
+
+	private Enemy Enemy { get; set; }
+	private Transform EnemyTransform { get; set; }
+	private Transform Vision { get; set; }
+	private Animator Animator { get; set; }
+
+	private void Awake() {
+		if (this.transform.parent.TryGetComponent(out Enemy enemy)) {
+			this.Enemy = enemy;
+		} else {
+			throw new Exception("Cant find Enemy script i parent.");
+		}
+		this.Animator = this.GetComponent<Animator>();
+	}
+	private void Start() {
+		this.InitializeEnemyBehaviour();
+	}
+
+	private void InitializeEnemyBehaviour() {
+
+		var type = Enemy.EnemyType;
+		this.PatrolSpeed = type.PatrolSpeed;
+		this.ChaseSpeed = type.ChaseSpeed;
+		this.VisionLength = type.VisionLength;
+
+		EnemyTransform = Enemy.transform;
+		this.Vision = Enemy.FrontPoint;
+
+		this.SetState(BehaviourState.IDLE);
+	}
+
+	private void FixedUpdate() {
+		var target = TryGetTarget();
+		switch (this.CurrentState) {
+			case BehaviourState.IDLE:
+				TryChangeToPatrol();
+				SetChaseIfCanSeeTarget(target);
+				break;
+			case BehaviourState.PATROL:
+				Move(this.PatrolSpeed);
+				TryChangeToIdle();
+				SetChaseIfCanSeeTarget(target);
+				break;
+			case BehaviourState.CHASE:
+				Move(this.ChaseSpeed);
+				SetIdleIfLostSightOfTarget(target);
+				SetAttackIfTargetIsInReach(target);
```

Comment:
Kan man ikke sjekke for `target==null` her, for kun å gjøre det aktuelle kallet? Så slipper man å sjekke igjen inni hver funksjon?